terraform/crates-io/impl/fastly-static.tf

```diff
@@ -0,0 +1,44 @@
+resource "fastly_service_vcl" "static" {
+  name = var.static_domain_name
+
+  domain {
+    name = var.static_domain_name
+  }
+
+  backend {
```

Comment:
We are using an `origin_group` on CloudFront with two S3 buckets and a failover strategy based on HTTP `5xx` responses. Achieving the same setup is slightly involved on Fastly:

- The `fastly_service_vcl` supports [failovers](https://developer.fastly.com/learning/concepts/failover/) through _automatic load-balancing_ or a _fallback director_. The latter is [not supported yet](https://github.com/fastly/terraform-provider-fastly/issues/436) in Terraform, and the first requires us to set up health checks against S3.
- We can probably write some custom VCL to replicate the functionality.
- The `fastly_service_compute` allows us to deploy a [Compute@Edge](https://github.com/fastly/compute-starter-kit-rust-static-content/blob/main/src/main.rs) function, which can mirror CloudFront's behavior. Requires building a local WASM module.

The benefit of `compute` over `vcl` seems to be that we have full control over the request cycle in a Rust runtime. The downside is that we are now relying on our own potentially buggy code instead of a provided Terraform module. If we do custom stuff, I'd still rather write some Rust than VCL.

Reply:
I feel like I commented somewhere but can't find it -- one alternative is to "just" have Fastly hit cloudfront as its origin, right? We can add a CNAME to cloudfront (e.g., cloudfront.static.crates.io) which Fastly hits for its origin.

Reply:
Another problem would be dealing with invalidations. If both CDNs are independent we can issue invalidations to both, while if Fastly points to CloudFront we would first need to invalidate CloudFront, wait for the invalidation to be finished, and then invalidate Fastly. That sounds like a mess to implement.

Reply:
If possible I'd prefer not to have a Rust binary handling this. It makes deploying changes to the CDN configuration more complex, and has a higher chance of bugs.

> and the first requires us to set up health checks against S3.

While it wouldn't be exactly the same behavior as CloudFront, setting up a health check would be trivial: we can create an empty[^1] `health` file into the S3 bucket and have Fastly use that as the health check. It would still cover the reason why the fallback is there (S3 is down in that region), even though it's not exactly the same mechanism as CloudFront.

[^1]: Well, in the content of the health file I'd put something like "needed by the Fastly CDN for the health check" so that future us will know why that file exists.

Reply:
Even with a health check, we will not be able to set up a primary/fallback configuration with Fastly. Fastly can only load balance between two buckets, and retry the other if the first request fails. Which is not a big issue in my opinion, but something to note. (See the remark about the _fallback director_ [not being supported](https://github.com/fastly/terraform-provider-fastly/issues/436) in Terraform.)

Personally, I don't have an issue using a compute function here. I think that we will need a binary for the Rust releases, since that does a lot of rewriting that cannot be expressed easily in VCL (as far as I can tell). So sooner or later, we have to learn how to manage it anyways.

But I'm fine leaving that discussion for later, and configuring `static.crates.io` with VCL for now.